src/tools/perf/perftest_mad.c

```diff
@@ -0,0 +1,694 @@
+                                      ib_portid_t *remote_port)
+{
+    int fd;
+    int ret;
+    void *umad;
+    uint8_t *data;
+    int len; /* cannot use 'size_t' here */
+    size_t ref_len;
+    struct ib_user_mad *user_mad;
+    ucs_status_t status;
+
+    fd      = mad_rpc_portid(rte_group->mad_port);
+    ref_len = *avail + IB_VENDOR_RANGE2_DATA_OFFS;
+    umad    = calloc(1, ref_len + umad_size());
+    if (umad == NULL) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+retry:
+    if (ref_len > (INT_MAX - umad_size())) {
+        status = UCS_ERR_INVALID_PARAM;
+        goto terminate;
+    }
+
+    user_mad = umad;
+    len      = ref_len;
+
+    ret = umad_recv(fd, umad, &len, 3 * 1000);
+    if (ret < 0) {
+        if ((errno == ETIMEDOUT) || (errno == EAGAIN)) {
+            status = UCS_ERR_NO_PROGRESS;
+            goto terminate;
+        }
+        if ((errno != EINVAL) && (errno != ENOSPC)) {
+            ucs_error("MAD: receive: failed, umad len:%d, ret:%d, errno:%d",
+                      len, ret, errno);
+            status = UCS_ERR_IO_ERROR;
+            goto terminate;
+        }
+        /* EINVAL: no length info, ENOSPC: needed length is set */
+        if (errno == EINVAL) {
+            ref_len *= 2; /* 'len' has become invalid */
+        } else {
+            ref_len = len;
+        }
+        umad = realloc_or_free(umad, umad_size() + ref_len);
+        if (umad == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+        goto retry;
+    }
+
+    if (perftest_mad_get_remote_port(umad, remote_port) != UCS_OK) {
+        status = UCS_ERR_IO_ERROR;
+        goto terminate;
+    }
+
+    len -= IB_VENDOR_RANGE2_DATA_OFFS;
+    if (len <= 0) {
+        if (user_mad->status == ETIMEDOUT) {
+            ucs_error("MAD: receive: Remote unreachable");
+            status = UCS_ERR_UNREACHABLE;
+        } else {
+            status = UCS_ERR_OUT_OF_RANGE;
+        }
+        goto terminate;
+    }
+
+    ret = umad_status(umad);
+    if (ret) {
+        ucs_warn("MAD: receive: Status failure: %d", ret);
+        status = UCS_ERR_REJECTED;
+        goto terminate;
+    }
+    status = UCS_OK;
+
+    data = umad_get_mad(umad);
+    data = UCS_PTR_BYTE_OFFSET(data, IB_VENDOR_RANGE2_DATA_OFFS);
+    if (perftest_mad_is_ping(data, len)) {
+        ucs_debug("PONG");
+        goto retry;
+    }
+
+    if (len > *avail) {
+        status = UCS_ERR_MESSAGE_TRUNCATED;
+        len    = *avail;
+    }
+    memcpy(buffer, data, len);
+    *avail = len;
+terminate:
+    free(umad);
+    return status;
+}
+
+static ucs_status_t
+perftest_mad_recv_from_remote(perftest_mad_rte_group_t *rte_group, void *buffer,
+                              size_t *avail, const ib_portid_t *target_port)
+{
+    ucs_status_t ret        = UCS_ERR_IO_ERROR;
+    ib_portid_t remote_port = {
+        .lid = 0
+    };
+    size_t size;
+
+    while ((remote_port.lid == 0) || (remote_port.lid != target_port->lid)) {
```

Comment:
Listening is datagram oriented, any remote trying to communicate on the perftest umad rte class would interfere with the current flow. As soon as we accept, we remember the initial lid and ignore any mad from different origin.

For instance it can be another perftest client started from another machine while a long run is already going.

Reply:
maybe convert it to for (;;) loop, and return from the function when found a valid response.
instead of using remote_port.lid being 0

Reply:
fixed